File: src/robot_driver/robot_driver/TurnLeftActionServer.py
import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer, GoalResponse, CancelResponse
from bot_behavior_interfaces.action import TurnLeft  # Import the TurnLeft action
from geometry_msgs.msg import Twist

class TurnLeftActionServer(Node):
    def __init__(self):
        super().__init__('turn_left_action_server')

        # Declare and set the turning speed 
        self.turn_speed = 0.5  

        self._action_server = ActionServer(
            self,
            TurnLeft,
            'turn_left',
            self.execute_callback,
            goal_callback=self.goal_callback,
            cancel_callback=self.cancel_callback)
        self.publisher_ = self.create_publisher(Twist, 'cmd_vel', 10)
        self.timer = None

    def goal_callback(self, goal_request):
        self.get_logger().info('Received request to turn left.')
        return GoalResponse.ACCEPT

    def cancel_callback(self, goal_handle):
        self.get_logger().info('Received cancel request. Stopping the robot.')
        self.stop_robot()
        return CancelResponse.ACCEPT

    def execute_callback(self, goal_handle):
        self.get_logger().info(f'Executing turn left with speed {self.turn_speed}...')
        
        twist_msg = Twist()
        twist_msg.angular.z = self.turn_speed  # Use the class attribute for angular speed

        # Create a timer to publish twist messages periodically
        self.timer = self.create_timer(0.1, self.publish_twist_msg)
        self.timer_msg = twist_msg

        while not goal_handle.is_cancel_requested:
            rclpy.spin_once(self, timeout_sec=0.1)

        if goal_handle.is_cancel_requested:
            self.get_logger().info('Goal canceled. Stopping the robot.')
            self.stop_robot()
            goal_handle.canceled()
            return TurnLeft.Result()

    def publish_twist_msg(self):
        if self.timer_msg:
            self.publisher_.publish(self.timer_msg)

    def stop_robot(self):
        # Publish zero velocity to stop the robot
        twist_msg = Twist()
        twist_msg.linear.x = 0.0
        twist_msg.angular.z = 0.0
        self.publisher_.publish(twist_msg)
        if self.timer:
            self.timer.cancel()

def main(args=None):
    rclpy.init(args=args)
    turn_left_action_server = TurnLeftActionServer()
    rclpy.spin(turn_left_action_server)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/robot_driver_cpp/src/robot_driver_cpp/turn_left_action_server.cpp
#include "robot_driver_cpp/turn_left_action_server.hpp"

using namespace std::chrono_literals;

TurnLeftActionServer::TurnLeftActionServer() : Node("turn_left_action_server")
{

    action_server_ = rclcpp_action::create_server<TurnLeft>(
        this,
        "turn_left",
        std::bind(&TurnLeftActionServer::handle_goal, this, std::placeholders::_1, std::placeholders::_2),
        std::bind(&TurnLeftActionServer::handle_cancel, this, std::placeholders::_1),
        std::bind(&TurnLeftActionServer::handle_accepted, this, std::placeholders::_1)
    );

    publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel", 10);
}

rclcpp_action::GoalResponse TurnLeftActionServer::handle_goal(
    const rclcpp_action::GoalUUID &uuid,
    std::shared_ptr<const TurnLeft::Goal> goal
)
{
    RCLCPP_INFO(this->get_logger(), "Received request to turn left.");
    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
}

rclcpp_action::CancelResponse TurnLeftActionServer::handle_cancel(
    const std::shared_ptr<GoalHandleTurnLeft> goal_handle
)
{
    RCLCPP_INFO(this->get_logger(), "Received cancel request. Stopping the robot.");
    stop_robot();
    return rclcpp_action::CancelResponse::ACCEPT;
}

void TurnLeftActionServer::handle_accepted(const std::shared_ptr<GoalHandleTurnLeft> goal_handle)
{
    std::thread{std::bind(&TurnLeftActionServer::execute, this, std::placeholders::_1), goal_handle}.detach();
}

void TurnLeftActionServer::execute(const std::shared_ptr<GoalHandleTurnLeft> goal_handle)
{
    auto goal = goal_handle->get_goal();
    turn_speed_ = goal->speed;
    RCLCPP_INFO(this->get_logger(), "Executing turn left with speed %f...", turn_speed_);
    
    twist_msg_.angular.z = turn_speed_;
    timer_ = this->create_wall_timer(
        100ms, [this]() { publish_twist_msg(); }
    );

    while (rclcpp::ok() && !goal_handle->is_canceling()) {
        std::this_thread::sleep_for(100ms);
    }

    if (goal_handle->is_canceling()) {
        RCLCPP_INFO(this->get_logger(), "Goal canceled. Stopping the robot.");
        stop_robot();
        goal_handle->canceled(std::make_shared<TurnLeft::Result>());
    } else {
        goal_handle->succeed(std::make_shared<TurnLeft::Result>());
    }
}

void TurnLeftActionServer::publish_twist_msg()
{
    if (publisher_) {
        publisher_->publish(twist_msg_);
    }
}

void TurnLeftActionServer::stop_robot()
{
    twist_msg_.linear.x = 0.0;
    twist_msg_.angular.z = 0.0;
    publisher_->publish(twist_msg_);
    if (timer_) {
        timer_->cancel();
    }
}
